### PR TITLE
Feature/map search

### PR DESCRIPTION
**Pull request by moreginger**
Improved search and usability in the map list. Specifically:

+ Filter field now searches map name, dimensions (wxh string), type and terrain
+ Filter is split into space separated terms and each term applied sequentially as AND (all terms must match). This is intended to be an intuitive way to use search.
+ Filter field has initial focus
+ Pressing enter selects the top map
+ Hint text always shown in focused controls when there is no user text
+ Fixed bug where the contents of the list weren't reset after filtering, closing and opening the list (because setText doesn't fire change event).

### CODE REVIEW

File: libs/chiliui/chili/headers/skinutils.lua [hunk 1]
```diff
@@ -613,7 +613,7 @@ function DrawEditBox(obj)
 	local font = _GetControlFont(obj)
 	local displayHint = false
 
-	if text == "" and not obj.state.focused then
+	if text == "" then
```
Comment (moreginger) [inline libs/chiliui/chili/headers/skinutils.lua:616]: I can see this being controversial, but I think it's better to keep the hint text until the user has typed something. This depends on the hint text having an obviously different format / syntax to what the user would type.

**Comment by moreginger** [comment]
Screenshot from 2021-02-14 20-57-32

**Comment by Licho1** [comment]
I would say that map search should optionally include non-downloaded maps. 
We should add a way to ask server for map list (via lobby protocol).

**Comment by moreginger** [comment]
> I would say that map search should optionally include non-downloaded maps.
> We should add a way to ask server for map list (via lobby protocol).

Sounds like something to work towards. What is "Download Maps" supposed to do currently? I get (on Linux):

```
Zero-K.exe Information: 0 : Chobbyla << OpenUrl {"Url":"https://zero-k.info/Maps?asmallcake=f304633d-b404-4c48-9dde-39f9146cdb46"}
Zero-K.exe Error: 0 : Error opening URL https://zero-k.info/Maps?asmallcake=f304633d-b404-4c48-9dde-39f9146cdb46 : System.ComponentModel.Win32Exception (0x80004005): Cannot find the specified file
```

**Comment by GoogleFrog** [comment]
What is this PR supposed to make the filter field do?

"Download maps" is a link to the ZK site map page.

**Comment by moreginger** [comment]
> What is this PR supposed to make the filter field do?

Sorry, commit messages were more descriptive. Have added summary.

**Comment by GoogleFrog** [comment]
Sounds good. I'll have a look on the weekend.

**Pull request merged**